IBN-2008 handle special case ft_value_1_byte

cloewen opened the pull request:
We have to introduce a special case here for the ft_value_1byte function. Currently i can´t get the smarthome project to work in my IDE, so lets test this fix on the target.

mkhl commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/ArithmeticGroupFunction.java:
```diff
@@ -537,27 +542,29 @@ public Threshold(State upperValue, State lowerValue, DecimalType upperLimit, Dec
+				boolean isLower = true;
+				for (Item item : items) {
+					DecimalType itemState = (DecimalType) item.getStateAs(DecimalType.class);
+					if (getFunctype(item).equals("ft_value_1_byte")) {
```
I think I’d like it better if we pushed the magic constant string down into the method.

A _concrete_ version of the method could return a boolean, and we’d call it like

```
if (usesPercentType(item)) {
    itemState = itemState.as(PercentType.class);
}
```

A more _generic_ version could instead return the State to use, like

```
Class<? extends DecimalType> type = findItemType(item);
DecimalType itemState = item.getStateAs(type);
```

(this would also replace the line above)

mkhl commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/ArithmeticGroupFunction.java:
```diff
@@ -537,27 +542,29 @@ public Threshold(State upperValue, State lowerValue, DecimalType upperLimit, Dec
+				for (Item item : items) {
+					DecimalType itemState = (DecimalType) item.getStateAs(DecimalType.class);
+					if (getFunctype(item).equals("ft_value_1_byte")) {
+						itemState.value = (itemState.value.multiply(new BigDecimal(100)));
```
Looking at DecimalType I believe this is identical to:

```
(PercentType) itemState = itemState.as(PercentType.class);
```

but see above

mkhl commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/ArithmeticGroupFunction.java:
```diff
@@ -573,6 +580,20 @@ public State getStateAs(Set<Item> items, Class<? extends State> stateClass) {
         }
+        
+		private String getFunctype(Item item) {
+			String memberFuncType = "Undef";
```
If we change this method to return boolean or Class, we wouldn’t need this magic constant. The default return values would be `false` and `DecimalType.class` respectively.

mkhl commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/ArithmeticGroupFunction.java:
```diff
@@ -573,6 +580,20 @@ public State getStateAs(Set<Item> items, Class<? extends State> stateClass) {
+				if (jsonElementTag instanceof JsonObject) {
+					JsonObject jsonObjectMemberTag = ((JsonObject) jsonElementTag);
+					if (jsonObjectMemberTag.has("functype")) {
+						memberFuncType = jsonObjectMemberTag.get("functype").toString();
```
Isn’t it simpler to use `….getAsString()` and not have to remove the surrounding `"`?

mkhl commented:
I took the liberty of addressing my own feedback. I haven’t tried it myself yet either.

mkhl committed:
IBN-2008: Threshold: Add optional factor parameter

When handling a child whose state should be interpreted as PercentType
(i.e. whose state is in [0..1] but represents a percentage in [0..100]),
we need to account for this factor when calculating the state of our function.

Since we cannot access the function type of the child item in the calculation,
we instead pass a factor as the optional fifth parameter to the function,
which it can then multiply the child item state with,
and which defaults to one.
